### PR TITLE
derive: Avoid emitting provided PartialEq, PartialOrd methods for c-like enums

### PR DESCRIPTION
**Pull request by bluss**
derive: Avoid emitting provided PartialEq, PartialOrd method for c-like enums

`ne` is completely symmetrical with the method `eq`, and we can save
rust code size and compilation time here if we only emit one of them
when possible.

One case where it's easy to recognize is when it's a C-like enum. Most
other cases can not omit ne, because any value field may have a custom
PartialEq implementation.

### CODE REVIEW

**Comment by bluss** [comment]
Using perf stat -r5 to measure compile time on a [small test crate with 140 ten-variant enums (source)](https://gist.github.com/bluss/2e47edf1b062b8cdc019).

- Before: 2.54 seconds time elapsed   ( +-  9,03% )  2170,409698      task-clock (msec)
- After: 1.55 seconds time elapsed   ( +-  9,60% )  1305,038461      task-clock (msec)

File: src/libsyntax_ext/deriving/cmp/partial_eq.rs [hunk 1]
```diff
@@ -11,13 +11,33 @@
...
 use syntax::parse::token::InternedString;
 use syntax::ptr::P;
 
+fn is_clike_enum(item: &Annotatable) -> bool {
+    match *item {
+        Annotatable::Item(ref item) => {
+            match item.node {
+                ItemKind::Enum(ref enum_def, _) => {
+                    enum_def.variants.iter().all(|v|
+                        if let VariantData::Unit(..) = v.node.data {
```
Comment (petrochenkov) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: `Unit` is not necessary, zero fields would be enough.

Comment (petrochenkov) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: I mean, `enum E { V {} }` isn't considered "C-like", but it's still amenable to this `derive` optimization.

**Comment by bluss** [comment]
Timing on src/libgraphviz (which has two enums that match this optimization)

libgraphviz debug build (perf stat -r20)

- before 0,90 seconds time elapsed ( +-  3,18% )
- after 0,76 seconds time elapsed   ( +-  2,67% )

libgraphviz -O build  (perf stat -r20)

- before 1,72 seconds time elapsed  ( +-  2,86% )
- after 1,46 seconds time elapsed  ( +-  4,07% )

as a control, librand (no enums match) debug build

- before 1,53 seconds time elapsed  ( +-  4,48% ) 
- after 1,49 seconds time elapsed   ( +-  3,64% )

File: src/libsyntax_ext/deriving/cmp/partial_eq.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: Also, `#[derive(PartialEq)]` for `struct S;`/`struct S {}` can be optimized too.

Comment (bluss) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: Good points. I'll try to incorporate that. Do I need to pretend zero field tuple variants can exist (they are disallowed)?

**Comment by bluss** [comment]
Part of issue  #31972

**Comment by brson** [comment]
r=me

Great little optimization.

File: src/libsyntax_ext/deriving/cmp/partial_eq.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: `derive` can just use `variant_data.fields().is_empty()`, caring about empty tuple structs is not its responsibility.

Comment (bluss) [inline src/libsyntax_ext/deriving/cmp/partial_eq.rs]: Thanks for the pointer. I usually don't poke around in these parts!

**Comment by durka** [comment]
Just for paranoia, did you confirm no runtime impact?

**Comment by bluss** [comment]
I didn't, I can try.

**Comment by bluss** [comment]
It's still the same. Testcase was just this microbenchmark though https://gist.github.com/bluss/ce593a7a75bd55896948

**Comment by bluss** [comment]
I did some tests using only the `partial_cmp` for a c-like enum versus the full ensemble `partial_cmp, lt, le, gt, ge`. It seems like there is no performance change.

So I added a change to treat `PartialOrd` the same way, it has four methods we can let have just the default implementation. It saves a lot of code for an enum of many variants. However, `PartialOrd` is not at all as common as `PartialEq`.

Since the code is general for "enum or struct has no fields", it also affects unit structs, which are relatively common.

**Comment by bluss** [comment]
Oh right thanks! I remember I didn't want to bother you with the PR during release rush.. two weeks ago.

**Pull request merged**